### PR TITLE
[Core] Modeler Stages

### PR DESCRIPTION
**Pull request by tteschemacher**
Dear all,

I attached the stages of the modeler, which I see important to cover all analyses types.

Please give me your feedback about it!?

Thanks

### CODE REVIEW

**Comment by tteschemacher** [comment]
@KratosMultiphysics/technical-committee 

Important question: Shall the Model be passed in every function or in the initialize of the modeler?

**Comment by philbucher** [comment]
> @KratosMultiphysics/technical-committee
> 
> Important question: Shall the Model be passed in every function or in the initialize of the modeler?

same as the processes
If it needs it later it needs to save it

My opinion

**Comment by tteschemacher** [comment]
Thank you for your comment @philbucher .

The concern that I have is that at modeler construction time the model parts are not yet created. However, by storing the model the problem can be ovoided. It is just a thing to think and take care about. 

The drawback by passing it every time is that the model might not be accessible at any time but only the model_part. This would be an argument for having the model passed in the beginning.

I'll be happy about any decision!

File: kratos/modeler/modeler.h [hunk 1]
```diff
@@ -68,85 +48,47 @@ class Modeler
...
-
+    virtual ~Modeler() = default;
 
     ///@}
-    ///@name Operators
+    ///@name Modeler Stages at Initialize
     ///@{
 
+    /// Import geometry models from external input.
+    virtual void ImportGeometryModel() const
```
Comment (pooyan-dadvand) [inline kratos/modeler/modeler.h]: These methods cannot be const

Comment (tteschemacher) [reply kratos/modeler/modeler.h]: sure. thanks for the comment. It is changed

File: kratos/modeler/modeler.h [hunk 2]
```diff
@@ -68,85 +48,53 @@ class Modeler
...
-
-
     ///@}
     ///@name Life Cycle
     ///@{
 
     /// Default constructor.
-    Modeler() {}
+    Modeler(
+        const Parameters ModelerParameters = Parameters())
```
Comment (philbucher) [inline kratos/modeler/modeler.h]:
```suggestion
        Parameters ModelerParameters = Parameters())
```

that cannot be const either

File: kratos/python/add_modeler_to_python.cpp [hunk 3]
```diff
@@ -69,15 +61,30 @@ void  AddModelerToPython(pybind11::module& m)
 {
     py::class_<Modeler, Modeler::Pointer>(m,"Modeler")
     .def(py::init<>())
-    .def("GenerateModelPart",&GenerateModelPart)
+    .def(py::init<const Parameters>())
```
Comment (philbucher) [inline kratos/python/add_modeler_to_python.cpp]:
```suggestion
    .def(py::init<Parameters>())
```

File: kratos/modeler/modeler.h [hunk 4]
```diff
@@ -68,85 +48,53 @@ class Modeler
...
+        : mParameters(ModelerParameters)
+    {}
+   
+    /// Constructor with Model
+    Modeler(
+        Model& rModel,
+        const Parameters ModelerParameters = Parameters())
```
Comment (philbucher) [inline kratos/modeler/modeler.h]:
```suggestion
        Parameters ModelerParameters = Parameters())
```

**Review by philbucher — COMMENTED** [review]
I think this needs one constructor taking `Model` and `Parameters`

File: kratos/modeler/modeler.h [hunk 5]
```diff
@@ -194,107 +132,22 @@ class Modeler
...
-    ///@name Un accessible methods
-    ///@{
-
-    /// Assignment operator.
-    Modeler& operator=(Modeler const& rOther);
-
-    /// Copy constructor.
-    Modeler(Modeler const& rOther);
-
+    const Parameters mParameters;
```
Comment (philbucher) [inline kratos/modeler/modeler.h]:
```suggestion
    Parameters mParameters;
```

**Comment by tteschemacher** [comment]
Thank you @philbucher for your comments

**Comment by tteschemacher** [comment]
> 
> 
> I think this needs one constructor taking `Model` and `Parameters`

True. It needs also a default constructor for the backwards compability.

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee after long discussion we think this  PR should be discussed inviting @tteschemacher and having @rubenzorrilla on board again. So we will postpone it to when @rubenzorrilla is back.

**Comment by tteschemacher** [comment]
@KratosMultiphysics/technical-committee I changed everything according to your suggestions.

**Review by philbucher — APPROVED** [review]

**Comment by tteschemacher** [comment]
Yuhuu. Thank you!

**Comment by pooyan-dadvand** [comment]
@tteschemacher After revising the final changes in the merged PR @KratosMultiphysics/technical-committee we are not sure if the names are coherent and seems to be a misunderstanding in those names:

from one side we have `SetupModelGeometry `and `PrepareModelGeometry `which seems that setup is for import and prepare is as before but for the modelpart we have `SetupModelPart `and `GenerateModelPart `in which steup seems to be a replace of prepare. 

Would you please clarify this?

**Comment by tteschemacher** [comment]
@pooyan-dadvand I understand the confusion. The 'GenerateModelPart' is there due to backwards compatibility of the Modeler class. Thus, it also doesn't follow the interface rules. However, it's not part of the Modeler stages. There we have the 3 discussed and agreed changes.

**Comment by pooyan-dadvand** [comment]
So I would change the `GenerateModelPart` into `PrepareModelPart`. @KratosMultiphysics/technical-committee?

**Pull request merged**